core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -1080,11 +1079,7 @@ class LogManager(logDirs: Seq[File],
 
         log
       }
-      // When running a ZK controller, we may get a log that does not have a topic ID. Assign it here.
-      if (log.topicId.isEmpty) {
-        topicId.foreach(log.assignTopicId)
-      }
-
+      
```

Comment:
Should we change the topid id so it's not optional?

Reply:
Although this method doesn't utilize Optional's functionality internally, the calling method `Partition#createLogIfNotExists` uses Optional features, and Optional is still required when passing to the `UnifiedLog` constructor later. Therefore, I don't see a compelling reason to avoid using Optional at this point.

https://github.com/apache/kafka/blob/78e35457ee966695bd392aa4e4c16cbfe919ad4f/core/src/main/scala/kafka/cluster/Partition.scala#L463
https://github.com/apache/kafka/blob/78e35457ee966695bd392aa4e4c16cbfe919ad4f/core/src/main/scala/kafka/log/UnifiedLog.scala#L191

Reply:
Makes sense. Unrelated to this PR, but a good clean-up would be to extract the topic of creation so it's always set in UnifiedLog (the way it works now unnecessarily exposes an optional topic id even though it is never really optional with kraft).

Reply:
It a good idea, I will file a Jira to trace this issue.

Reply:
we should do this cleanup after all zk-related paths are removed from `ReplicaManager`